covenants/src/protocol_tests.rs

```diff
@@ -0,0 +1,403 @@
+#[cfg(test)]
+mod tests {
+    use crate::{Borrower0, Lender0};
+    use anyhow::Result;
+    use bitcoin_hashes::Hash;
+    use elements::{
+        bitcoin::Amount,
+        hashes::hash160,
+        opcodes,
+        script::Builder,
+        secp256k1::{rand::thread_rng, SecretKey, SECP256K1},
+        sighash::SigHashCache,
+        AssetId, Script, SigHashType, TxIn,
+    };
+    use elements_harness::{elementd_rpc::ElementsRpc, Elementsd};
+    use secp256k1_zkp::Message;
+    use std::env;
+    use testcontainers::clients::Cli;
+
+    #[tokio::test]
+    async fn borrow_and_repay() {
+        // force enabling log output
+        // TODO: remove me again
+        env::set_var("RUST_LOG", "DEBUG");
+        let _ = env_logger::builder().is_test(true).try_init();
+
+        let tc_client = Cli::default();
+        let (client, _container) = {
+            let blockchain = Elementsd::new(&tc_client, "0.18.1.9").unwrap();
+
+            (
+                elements_harness::Client::new(blockchain.node_url.clone().into_string()).unwrap(),
+                blockchain,
+            )
+        };
+
+        let master_blinding_key = client.dumpmasterblindingkey().await.unwrap();
+        let master_blinding_key = hex::decode(master_blinding_key).unwrap();
+
+        let bitcoin_asset_id = client.get_bitcoin_asset_id().await.unwrap();
+        let usdt_asset_id = client.issueasset(40.0, 0.0, false).await.unwrap().asset;
+
+        let address = client
+            .get_new_address(Some("blech32".into()))
+            .await
+            .unwrap();
+        client
+            .send_asset_to_address(&address, Amount::from_btc(5.0).unwrap(), None)
+            .await
+            .unwrap();
+        let miner_address = client
+            .get_new_address(Some("blech32".into()))
+            .await
+            .unwrap();
+        client.generatetoaddress(10, &miner_address).await.unwrap();
+
+        let (borrower, _borrower_address) = {
+            let collateral_amount = Amount::ONE_BTC;
+            let collateral_inputs = find_inputs(&client, bitcoin_asset_id, collateral_amount * 2)
+                .await
+                .unwrap();
+            let address = client
+                .get_new_address(Some("blech32".into()))
+                .await
+                .unwrap();
+            let address_blinding_sk =
+                derive_blinding_key(master_blinding_key.clone(), address.script_pubkey()).unwrap();
+
+            let timelock = 0;
+
+            let borrower = Borrower0::new(
+                address.clone(),
+                address_blinding_sk,
+                collateral_amount,
+                collateral_inputs,
+                Amount::ONE_SAT,
+                timelock,
+                bitcoin_asset_id,
+                usdt_asset_id,
+            )
+            .unwrap();
+
+            (borrower, address)
+        };
+
+        // TODO: Use a separate wallet per actor. Using the same wallet is confusing and bug-prone.
```

Comment:
What do you think @luckysori : shall we tackle this before merging? 
Could be easier to find potential flaws now than later.